### PR TITLE
enemySpawner init

### PR DESCRIPTION
**Pull request by JeyR01**

### CODE REVIEW

File: OPCB/functions/missions/El_Chapo.sqf
Comment (KHunter-Arma) [inline OPCB/functions/missions/El_Chapo.sqf:1]: I don't understand what your doing here in L11 and L20. You're creating an officer twice?

File: OPCB/functions/missions/gdrunken_spawn.sqf [hunk 1]
```diff
@@ -1,38 +1,37 @@
...
+private _tanks = [];
 
 for "_i" from 0 to 3 do {
-	_suitable = [0, 0, 0];
+	_suitable = [0, 0, 0, 0];
 	_maxDist = 1500;
 	_list = [];
 
-	while {((count _suitable) == 3) || {(count _list) > 0}} do {
+	while {count _suitable == 4 || {surfaceIsWater _suitable}} do {
```
Comment (KHunter-Arma) [inline OPCB/functions/missions/gdrunken_spawn.sqf:10]: Why did you change this to 4? If it fails, it will be a 3 element array, which will default to the map centre, which may or may not be on water. If it succeeds it will be 2. This is wrong.

File: OPCB/functions/missions/gdrunken_spawn.sqf [hunk 2]
```diff
@@ -1,38 +1,37 @@
...
 		
-		if ((count _suitable) == 3) then {
-			_maxDist = worldSize min (_maxDist + 100);
-		} else {
-			_list = nearestTerrainObjects [_suitable,["TREE","BUILDING","RUIN","ROCK","HOUSE"], 55,false];
-		};
+		_maxDist = worldSize min (_maxDist + 100);
+		
+	
+		_list = nearestTerrainObjects [_suitable,[], 55,false];
```
Comment (KHunter-Arma) [inline OPCB/functions/missions/gdrunken_spawn.sqf:17]: If you're not making use of `list` within the loop, then take it out of the loop.

File: OPCB/initPlayerLocal.sqf [hunk 3]
```diff
@@ -79,9 +54,35 @@ jeff addaction ["Lights off", {
...
+	waitUntil{
+		sleep 1;
+		_initDone = player getVariable ["InitDone", false];
+		_initDone
+	};
+
+	#include "data\whitelistedUnitTypes.sqf";
+
+	if (typeof player in _whitelistedUnitTypes) then {
+		_whiteListed = player getVariable ["WhiteListed", false];
```
Comment (KHunter-Arma) [inline OPCB/initPlayerLocal.sqf:74]: Are these variables going to persist after respawn?

**Review by KHunter-Arma — CHANGES_REQUESTED** [review]

File: OPCB/functions/BADCO_Arsenal.sqf
Comment (KHunter-Arma) [inline OPCB/functions/BADCO_Arsenal.sqf:1]: Can you revert this change? It's already done on the arsenal update branch and there's no need to create a conflict.

File: OPCB/Hz_pers_customLoadFunction.sqf [hunk 4]
```diff
@@ -43,23 +43,19 @@ publicVariable "OPCB_econ_currentTier";
...
+				MaxAttackHelis = MaxAttackHelis - 1;
+				publicVariable "MaxAttackHelis";
 			}];
 				
 		} else {
 			MaxTransHelis = MaxTransHelis + 1;
 			publicVariable "MaxTransHelis";
 
-			_vehicle addMPEventHandler ["MPKilled",
+			_vehicle addEventHandler ["Killed",
```
[the remark this answers is not in the document]
Comment (JeyR01) [reply OPCB/Hz_pers_customLoadFunction.sqf:55]: there is no locality. What client side function even calls this customLoadFunction?

File: OPCB/functions/tankspawner/spawn_tank_server.sqf [hunk 5]
```diff
@@ -20,11 +20,9 @@ if (_isAttack == 1) then {
 	
 	_tank call Hz_pers_API_addVehicle;
 	
-	_tank addMPEventHandler ["MPKilled",{
```
[the remark this answers is not in the document]
Comment (JeyR01) [reply OPCB/functions/tankspawner/spawn_tank_server.sqf:23]: spawn_tank_server is run on server side. No need to braodcast event handler.

File: OPCB/functions/helispawner/spawn_helicopter_server.sqf [hunk 6]
```diff
@@ -28,11 +28,9 @@ if (_isAttack == 1) then {
 	{ _helicopter removeWeaponGlobal getText (configFile >> "CfgMagazines" >> _x >> "pylonWeapon") } forEach getPylonMagazines _helicopter;
 	{ _helicopter setPylonLoadOut [_forEachIndex + 1, _x, true, _pylonPaths select _forEachIndex] } forEach _pylons;
 
-	_helicopter addMPEventHandler ["MPKilled",{ 
-		if (isServer) then {
-			MaxAttackHelis = MaxAttackHelis - 1;
-			publicVariable "MaxAttackHelis";
-		};
+	_helicopter addEventHandler ["Killed",{
```
[the remark this answers is not in the document]
Comment (JeyR01) [reply OPCB/functions/helispawner/spawn_helicopter_server.sqf:31]: spawn_helicopter_server is run on server side. No need to braodcast event handler.

File: OPCB/economy/fnc/OPCB_crateSpawner_fnc_spawnCrate_server.sqf [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+		if ((_x select 0) == _crateType) exitWith {
+			_cargoRequirement = _x select 1;
+		};
+	} foreach OPCB_econ_vehicleCargoSizes;
+	
+	[_crate, _cargoRequirement] call ace_cargo_fnc_setSize;
+	
+	_crate call Hz_pers_API_addCrate;
+	
+	_crate addEventHandler ["Killed",
```
[the remark this answers is not in the document]
Comment (JeyR01) [reply OPCB/economy/fnc/OPCB_crateSpawner_fnc_spawnCrate_server.sqf:31]: same for the previous 2 comments.

File: OPCB/economy/fnc/OPCB_crateSpawner_fnc_spawnCrate_server.sqf [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+	
+	_crate = _crateType createVehicle (_tankpos);	
+	_crate setdir 40;
+	
+	_cargoRequirement = 0;
+	{
+		if ((_x select 0) == _crateType) exitWith {
+			_cargoRequirement = _x select 1;
+		};
+	} foreach OPCB_econ_vehicleCargoSizes;
```
[the remark this answers is not in the document]
Comment (JeyR01) [reply OPCB/economy/fnc/OPCB_crateSpawner_fnc_spawnCrate_server.sqf:25]: good point

File: OPCB/functions/helispawner/spawn_drone_vehicle.sqf [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+	_tier = ["DRONE", _vehicle] call OPCB_econ_fnc_getVehicleTier;
+	_cost = ["DRONE", _tier] call OPCB_econ_fnc_getTierCost;
+
+	if (OPCB_econ_credits < _cost) exitWith {
+		hint "You don't have enough credits to buy this vehicle!";
+	};
+	OPCB_econ_credits = OPCB_econ_credits - _cost;
+	publicVariable "OPCB_econ_credits";
+
+	_drone = _vehicle createVehicle (_tankpos);
```
[the remark this answers is not in the document]
Comment (JeyR01) [reply OPCB/functions/helispawner/spawn_drone_vehicle.sqf:23]: currently this works. I don't want to touch it as long as it works.

File: OPCB/functions/missions/El_Chapo.sqf
Comment (JeyR01) [reply OPCB/functions/missions/El_Chapo.sqf:1]: honestly we should just remove this mission. It's only spawnable from the admin console anyway. This is giving me constant headaches

**Comment by KHunter-Arma** [comment]
@JeyR01 you need to read up on locality and vehicles. https://community.bistudio.com/wiki/Multiplayer_Scripting#Locality_changes
Especially pay attention to this line: **a driven vehicle is always local to the machine of its driver (not the commander's, not the gunner's)**

File: OPCB/functions/missions/gdrunken_spawn.sqf (hunk 1 above)
Comment (JeyR01) [reply OPCB/functions/missions/gdrunken_spawn.sqf:10]: this is from my previous tests. This works, count 3 also works, doesn't matter honestly. 
quote: Result Array -  in format [x,y] on success. When position cannot be found at all, default map center position is returned, which will be in format [x,y,0]. 
Which means that if it fails, it's count == 3 and runs again, so it doesn't matter if it's 4 or 3 dim for suitable locations.

File: OPCB/functions/missions/gdrunken_spawn.sqf (hunk 2 above)
Comment (JeyR01) [reply OPCB/functions/missions/gdrunken_spawn.sqf:17]: fair point

File: OPCB/initPlayerLocal.sqf (hunk 3 above)
Comment (JeyR01) [reply OPCB/initPlayerLocal.sqf:74]: yes I tried.

File: OPCB/functions/BADCO_Arsenal.sqf
Comment (JeyR01) [reply OPCB/functions/BADCO_Arsenal.sqf:1]: yes

**Comment by JeyR01** [comment]
> @JeyR01 you need to read up on locality and vehicles. https://community.bistudio.com/wiki/Multiplayer_Scripting#General_information_about_locality Especially pay attention to this line: **a driven vehicle is always local to the machine of its driver (not the commander's, not the gunner's)**

A vehicle might be local, but an eventhandler is not . Eventhandlers are local where you declare them.

**Comment by KHunter-Arma** [comment]
Exactly, and they also fire only when the vehicle it's attached to is local.

File: OPCB/functions/vehicleDeleteCheck.sqf [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+	params ["_veh", "_engineState"];
+	if (!_engineState && {alive _veh} && {!canMove _veh}) then {
+		_veh setVariable ["CHB_CanDespawn", true, false];
+	};
+}];
+
+//check if vehicle has been abandoned
+_veh addEventHandler ["GetOut",{
+	params ["_veh"];
+	if ((count crew _veh) == 0) then {
```
[the remark this answers is not in the document]
Comment (JeyR01) [reply OPCB/functions/vehicleDeleteCheck.sqf:51]: It was working the 1 time i was forcing it to happen. I haven't see a vehicle get hit and the crew survives.

File: OPCB/functions/missions/roadblock.sqf [hunk 11]
```diff
@@ -20,7 +24,7 @@ for "_i" from 0 to _roadblocks - 1 do
 			_roadblock = [_block,getpos _road, [0,0,0], _direction, true, true ] call LARs_fnc_spawnComp;
 			_spawnComp pushBack _roadblock;
 
-			_group = [getpos _road, east, selectRandom OPCB_unitTypes_grp_inf] call BIS_fnc_spawnGroup;
+			_group = [getpos _road, east, selectRandom OPCB_InfantryGroups_Insurgents] call BIS_fnc_spawnGroup;
```
[the remark this answers is not in the document]
Comment (JeyR01) [reply OPCB/functions/missions/roadblock.sqf:27]: yup

File: OPCB/functions/missions/roadblock_ins.sqf [hunk 12]
```diff
@@ -20,7 +24,7 @@ for "_i" from 0 to _roadblocks - 1 do
 			_roadblock = [_block,getpos _road, [0,0,0], _direction, true, true ] call LARs_fnc_spawnComp;
 			_spawnComp pushBack _roadblock;
 
-			_group = [getpos _road, east,selectRandom OPCB_unitTypes_grp_ins_inf] call BIS_fnc_spawnGroup;
+			_group = [getpos _road, east,selectRandom OPCB_InfantryGroups_OPFOR] call BIS_fnc_spawnGroup;
```
[the remark this answers is not in the document]
Comment (JeyR01) [reply OPCB/functions/missions/roadblock_ins.sqf:27]: yup

File: OPCB/functions/missions/gdrunken_spawn.sqf (hunk 1 above)
Comment (KHunter-Arma) [reply OPCB/functions/missions/gdrunken_spawn.sqf:10]: What? How does it run again if it's count==3 with this code?

File: OPCB/functions/missions/Secure.sqf [hunk 13]
```diff
@@ -1,24 +1,18 @@
...
-_spawncomps = [_guard] call CHAB_fnc_roadblock_rus;
-_comp = [_taskcomp,_guardpos, [0,0,0], random 360, true, true ] call LARs_fnc_spawnComp;
+[_base,east] call CHAB_fnc_enemySpawner;
 
-[_guard,10,1,3] call CHAB_fnc_spawn_rus;
-
-deleteVehicle _guard;
-
-_truckpos = _guardpos findEmptyPosition [1, 20, "O_Truck_03_device_F"];
+_truckpos = _base findEmptyPosition [1, 20, "O_Truck_03_device_F"];
```
[the remark this answers is not in the document]
Comment (JeyR01) [reply OPCB/functions/missions/Secure.sqf:15]: findSpot will clear and find an area that can hold the compound, then the vehicle can be safely place since the compound has a large enough space for the vehicle

File: OPCB/functions/missions/GDrunken.sqf [hunk 14]
```diff
@@ -1,36 +1,32 @@
...
+_comp = [_taskcomp,_base, [0,0,0], random 360, true, true ] call LARs_fnc_spawnComp;
 _reference = [ _comp ] call LARs_fnc_getCompObjects;
 
 _trucks = [];
 private _tanktype = "";
 for "_i" from 0 to 3 do {
-	_tanktype = selectRandom OPCB_unitTypes_veh_ins_armor;
-	_truckpos = _guardpos findEmptyPosition [1, 20, _tanktype];
+	_tanktype = selectRandom OPCB_ArmoredVehicles_OPFOR;
+	_truckpos = _base findEmptyPosition [1, 20, _tanktype];
```
[the remark this answers is not in the document]
Comment (JeyR01) [reply OPCB/functions/missions/GDrunken.sqf:17]: same as the secure task. the area will be cleared for the compound then the vehicle can spawn inside.

File: OPCB/functions/findSpot.sqf [hunk 15]
```diff
@@ -1,39 +1,40 @@
...
-			if ((count _suitable) == 3) then {
-				_startingDist = worldSize min (_startingDist + 500);
-			} else {
-				_list = nearestTerrainObjects [_suitable,["TREE","BUILDING","RUIN","ROCK","HOUSE"], _radius,false];
-			};		
+		while {count _suitable == 4 || {surfaceIsWater _suitable}} do {
 			
+			_suitable = [_basePos, 2000, _startingDist, 10, 0, 0.3, 0, [], [[0,0,0,0],[0,0,0,0]]] call BIS_fnc_findSafePos;
+
+			_list = nearestTerrainObjects [_suitable,[], _radius,false];
```
[the remark this answers is not in the document]
Comment (JeyR01) [reply OPCB/functions/findSpot.sqf]: yup, already taken out

File: OPCB/functions/helispawner/spawn_drone.sqf [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+
+_classname = _ctrl lbData 0;
+_picture = getText (configFile >> "CfgVehicles" >> _classname >> "editorPreview"); 
+_imageCtrl ctrlSetText _picture;
+
+_ctrl ctrlAddEventHandler ["LBSelChanged",{
+	params ["_control", "_selectedIndex"];
+	
+	_classname = _control lbData _selectedIndex;
+	_picture = getText (configFile >> "CfgVehicles" >> _classname >> "editorPreview");
```
[the remark this answers is not in the document]
Comment (JeyR01) [reply OPCB/functions/helispawner/spawn_drone.sqf:35]: it doesn't spawn any pictures, you are correct, and I was too busy fixing other stuff to take care of this issue. Currently this works and spawns the drones, that's all for now :D

File: OPCB/functions/missions/gdrunken_spawn.sqf (hunk 1 above)
Comment (JeyR01) [reply OPCB/functions/missions/gdrunken_spawn.sqf:10]: because I checked and all of the map default positions are water, so the 2nd check starts the loop again

File: OPCB/functions/vehicleDeleteCheck.sqf (hunk 10 above)
[the remark this answers is not in the document]
Comment (JeyR01) [reply OPCB/functions/vehicleDeleteCheck.sqf]: you are looking at an outdated version again maybe? I put a spawn in it. 
This GetOut EH is only needed if the vehicle drives into water. Then all the crew gets out. If the vehicle gets hit then the Engine EH fires and does the same thing. It's all tested

File: OPCB/functions/missions/gdrunken_spawn.sqf (hunk 1 above)
Comment (KHunter-Arma) [reply OPCB/functions/missions/gdrunken_spawn.sqf:10]: Ok but what about other maps?!

Comment (JeyR01) [reply OPCB/functions/missions/gdrunken_spawn.sqf:10]: I checked the maps in the repo folder. Future maps might differ.

Comment (KHunter-Arma) [reply OPCB/functions/missions/gdrunken_spawn.sqf:10]: This is going to end up being a miserable tech debt in the future that I hope you remember.

File: OPCB/functions/vehicleDeleteCheck.sqf (hunk 10 above)
[the remark this answers is not in the document]
Comment (KHunter-Arma) [reply OPCB/functions/vehicleDeleteCheck.sqf]: goddamn git

**Review by KHunter-Arma — APPROVED** [review]
